thin-edge/thin-edge.io
# [CIT-659] restart device local operation

## Proposed changes

We can now restart the device by publishing on the topic `tedge/commands/req/control/restart` with the id of the operation.

1. The previous code was reused as much as possible. Structs like ~~`SoftwareOperationStatus`~~ became `OperationStatus`. 
2. New request/response topics were created for this operation.
3. ~~`fail_pending_operations`~~ became `process_pending_operation`. This is the function that is executed when we start `tedge_agent`. This function now verifies if the restart was successful - making the rename appropriate. 
4. `StateStatus` is the struct that holds all possible operations `{Software, Restart, UnknownOperation}`. Where, `Software` made up of the __enum__ `SoftwareOperationVariants::{List, Update}` and `Restart` is made up of the __enum__ `RestartOperationStatus::{Pending, Restarting}`.
5. The restart operation functions are separated as a module. 

The restart logic goes as follows: 
1. receive restart message etc...
2. set operation status to `RestartOperationStatus::Pending`
3. create `/run/tedge_agent_restart` with current unix timestamp (in UTC)
4.  set operation status to `RestartOperationStatus::Restarting`
3.  Call `sync` command __then__ call `init 6` to reboot.
4. after reboot, the agent restarts calling the `process_pending_operation` where we __check__ if the reboot was a success.
5. if `/run/tedge_agent_restart` does __NOT__ exist, we conclude this was removed by the filesystem on reboot
6. if `/run/tede_agent_restart` __exists__, we compare the first value of `/proc/uptime` (seconds the system has been on) with the unix timestamp inside `/run/tedge_agent_restart` to decide if the restart operation was successful. 

## How to test
NOTE: this will not restart your device. The `init 6` command is currently changed to `echo 6` :)
After building and stopping existing agent (you might need to remove the lockfile in /run/lock/):
```
sudo ./target/debug/tedge_agent
```

```
mosquitto_sub -t '#' -v
```

send a restart operation:
```
mosquitto_pub -t 'tedge/commands/req/control/restart' -m '{"id":"1234"}'
```

Stop ./target/debug/tedge_agent and run again. Bullet points 4,5,6 (in the section above) detail the restart check, but to summarise again: 

1. if your computer has only been turned on in the last 5 minutes the reboot will pass (MAX_ALLOWED_DURATION in the code).
2. if you manually delete the /run/tedge_agent_restart file before running the `tedge_agent` component, the reboot will pass 

## Out of scope of this ticket
Also, I took the liberty to remove `chrono` as a dependency completely from the `tedge_agent` component - I hope you do not mind. It was a minor change, only in one file :) 


## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)

## Paste Link to the issue
https://cumulocity.atlassian.net/browse/CIT-659

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
